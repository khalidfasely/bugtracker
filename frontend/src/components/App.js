function App() {
  return (
    <div className="App">
      <header>
        Hello world from React        
      </header>
    </div>
  );
}

export default App;
