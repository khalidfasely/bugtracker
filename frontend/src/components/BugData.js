const BugData = ({ bug }) => (
    <div>
        <p>{bug?.id}</p>
        <p>{bug?.user?.username}</p>
        <p>{bug?.title}</p>
        <p>{bug?.description}</p>
        <p>{bug?.time}</p>
        <p>{bug?.classification}</p>
        <p>{`${bug?.active}`}</p>
        <div>
            {bug?.admins?.map(admin => <span key={admin} >{admin} </span>)}
        </div>
        <br />
        <div>
            {bug?.users_with?.map(user => <span key={user} >{user} </span>)}
        </div>
    </div>
);

export default BugData;