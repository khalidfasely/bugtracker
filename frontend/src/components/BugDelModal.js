import Modal from "react-modal";

const BugDelModal = ({ delModalOpen, setDelModalOpen, deleteBug }) => (
    <Modal
        isOpen={delModalOpen}
        onRequestClose={() => setDelModalOpen(false)}
    >
        If you delete this bug there is no way to return it!
        <button onClick={deleteBug} data-testid='delete_button'>Delete</button>
    </Modal>
);

export default BugDelModal;