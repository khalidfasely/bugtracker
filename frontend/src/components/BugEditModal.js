import Modal from "react-modal";

import NewBug from "./NewBug";

const BugEditModal = ({
    editModalOpen, setEditModalOpen,
    users, bug
}) => (
    <Modal
        isOpen={editModalOpen}
        onRequestClose={() => setEditModalOpen(false)}
    >
        <NewBug users={users} isEdit={true} bug={bug} setEditModalOpen={setEditModalOpen} />
    </Modal>
);

export default BugEditModal;