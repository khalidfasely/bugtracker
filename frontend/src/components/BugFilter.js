import { connect } from "react-redux";
import { setTextFilter, sortByDefault, sortByPriority } from "../actions/bugFilter";

export const BugFilter = ({ filters, setTextFilter, sortByDefault, sortByPriority }) => {
    const onSortByChange = (e) => {
        if (e.target.value === 'priority') {
            sortByPriority();
            return;
        };
        sortByDefault();
    };

    return (
        <div className="bugs-filter">
            <div className="input-group__item">
                <input
                    data-testid='search_input'
                    type="text"
                    className="text-input"
                    placeholder="Search bugs"
                    value={filters.text}
                    maxLength={35}
                    onChange={(e) => setTextFilter(e.target.value)}
                />
            </div>
            <div className="input-group__item">
                <select
                    data-testid='select_input'
                    className="select"
                    value={filters.sortBy}
                    onChange={onSortByChange}
                >
                    <option value="">Date</option>
                    <option value="priority">Priority</option>
                </select>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => ({
    filters: state.bugFilter
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByPriority: () => dispatch(sortByPriority()),
    sortByDefault: () => dispatch(sortByDefault())
});

export default connect(mapStateToProps, mapDispatchToProps)(BugFilter);