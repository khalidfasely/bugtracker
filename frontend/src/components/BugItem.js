import { Link } from "react-router-dom";
import { connect } from "react-redux";

import Comment from "./Comment";

const BugItem = ({uname, bug, comments}) => {
    console.log(bug, comments);
    return (
        <>
            {
                bug?.users_with?.includes(uname) ?
                <div>
                    **** bug <br /><h1>{bug.title}</h1>
                    **** comments {comments?.map(comment => <Comment comment={comment} />)}
                </div> :
                <div>Not able to see this bug! <Link to="/new">Create one!</Link></div>
            }
        </>
    );
}

const mapStateToProps = (state) => ({
    uname: state.auth.uname,
    bug: state.bugItem.bug,
    comments: state.bugItem.comments
});

export default connect(mapStateToProps)(BugItem);