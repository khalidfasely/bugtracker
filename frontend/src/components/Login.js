import { useState } from 'react';
import { useNavigate, Link } from 'react-router-dom';

const Login = () => {
    const history = useNavigate();
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');
    const [error, setError] = useState('');

    const login = async (e) => {
        e.preventDefault();
        setError('');

        // Check all field
        if (!(username && password)) {
            setError('Must set all fields!');
            return;
        };
    
        const res = await fetch('/api/login', {
          method: "POST",
          body: JSON.stringify({
            username,
            password
          })
        });
    
        const data = await res.json();

        if (!data.user) {
            setError(data.message);
            return;
        }
    
        console.log(data);

        history('/');
    };

    return (
        <div>
            <form onSubmit={login}>
            {error && <p>{error}</p>}
                <input
                    type='text'
                    placeholder='username'
                    value={username}
                    onChange={(e) => setUsername(e.target.value)}
                />
                <input
                    type='password'
                    placeholder='password'
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}
                />
                <button>Login</button>
            </form>
            <p>Haven't an account? Create one <Link to='/register'>here</Link>.</p>
        </div>
    );
};

export default Login;