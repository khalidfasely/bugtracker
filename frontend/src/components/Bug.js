import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { useParams } from "react-router-dom";
import { startSetBug } from "../actions/bug";

import BugItem from './BugItem';

export const Bug = ({ startSetBug }) => {
    const params = useParams();

    const [ableToSee, setAbleToSee] = useState(false);
    const [notFound, setNotFound] = useState('');

    useEffect(() => {
        startSetBug(params.pid, params.bid)
        .then((result) => {
            setNotFound('');

            if (result.message) {
                setNotFound(result.message);
            };
        })
        .then(() => setAbleToSee(true))
        .catch(er => console.error(er));
    }, [params])


    if (notFound) {
        return <div>Bug Not Found! -- {notFound}</div>;
    };

    return (
        <div>
            {
                ableToSee ?
                <div><BugItem /></div> :
                <div>You're not able to see this bug</div>
            }
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startSetBug: (pid, bid) => dispatch(startSetBug(pid, bid))
});

export default connect(undefined, mapDispatchToProps)(Bug);