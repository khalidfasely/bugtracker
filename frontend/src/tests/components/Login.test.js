import React from "react";
import { render, fireEvent, act } from "@testing-library/react";

import { Login } from "../../components/Login";

let startLogin;

beforeEach(() => {
    startLogin = jest.fn(() => Promise.resolve({}));
});

test('Should render Login component correctly', () => {
    const { asFragment } = render(<Login />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render Login button correctly', () => {
    const { getByTestId } = render(<Login startLogin={startLogin} />);
    const buttonEl = getByTestId("login_button");
    expect(buttonEl.textContent).toBe('Login');
});

test('Should change username input value', () => {
    const { getByTestId } = render(<Login startLogin={startLogin} />);
    const usernameInputEl = getByTestId("username_input");

    expect(usernameInputEl.value).toBe('');

    fireEvent.change(usernameInputEl, {
        target: {
            value: 'Username'
        }
    });

    expect(usernameInputEl.value).toBe('Username');
});

test('Should change password input value', () => {
    const { getByTestId } = render(<Login startLogin={startLogin} />);
    const passwordInputEl = getByTestId("password_input");

    expect(passwordInputEl.value).toBe('');

    fireEvent.change(passwordInputEl, {
        target: {
            value: 'Password'
        }
    });

    expect(passwordInputEl.value).toBe('Password');
});

test('Should not submit form with empty inputs', () => {
    const { getByTestId } = render(<Login startLogin={startLogin} />);
    const formEl = getByTestId("login_form");

    fireEvent.submit(formEl, {
        preventDefault: () => {}
    });

    const errorEl = getByTestId("error_message");
    expect(errorEl.textContent).toBe('Must set all fields!');
});

test('Should submit form with full inputs', async () => {
    const promise = Promise.resolve({});
    const startLoginIn = jest.fn(() => promise);

    const { getByTestId } = render(<Login startLogin={startLoginIn} />);
    const formEl = getByTestId("login_form");

    const passwordInputEl = getByTestId("password_input");
    fireEvent.change(passwordInputEl, {
        target: {
            value: 'Password'
        }
    });

    const usernameInputEl = getByTestId("username_input");
    fireEvent.change(usernameInputEl, {
        target: {
            value: 'Username'
        }
    });

    expect(startLoginIn).toHaveBeenCalledTimes(0);

    fireEvent.submit(formEl, {
        preventDefault: () => {}
    });

    await act(() => promise);
    expect(startLoginIn).toHaveBeenCalledTimes(1);
});