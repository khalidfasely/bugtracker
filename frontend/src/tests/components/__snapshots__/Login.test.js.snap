// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Login component correctly 1`] = `
<DocumentFragment>
  <div
    class="login-container"
  >
    <div
      class="login"
    >
      <form
        class="login-form"
        data-testid="login_form"
      >
        <input
          data-testid="username_input"
          placeholder="Username"
          type="text"
          value=""
        />
        <input
          data-testid="password_input"
          placeholder="Password"
          type="password"
          value=""
        />
        <button
          class="show-pswrd"
          disabled=""
          type="button"
        >
          Show password
        </button>
        <button
          class="login__button"
          data-testid="login_button"
        >
          Login
        </button>
      </form>
      <p>
        Haven't an account? Create one 
        <a>
          here
        </a>
        .
      </p>
    </div>
  </div>
</DocumentFragment>
`;
